Tetris branch

bberastain opened the pull request:
Trying to make an array of arrays then concatenate their strings at the end. Started with the board and worked out how to place one part, but I'm not yet sure how to approach the bigger pieces. The way I'll be able to learn the quickest is to ask better questions, but I don't yet know how to phrase these questions. I realize by the way you structured the other pieces that there's formatting to arrays I don't know about, but I don't know what to search in order to learn more.

bberastain committed:
Add files via upload

bberastain committed:
Delete tetris_test.py

bberastain committed:
Add files via upload

bberastain committed:
Update array_test.py

bberastain committed:
Add files via upload

fmezas commented:
![image](https://cloud.githubusercontent.com/assets/540110/23827312/b82f3dae-0665-11e7-9d0c-e589e2d449b9.png)

the image shows my editor complaining about formatting (and maybe some issues detected in static analysis) in your python code (that's all the underlining in red). given that spaces are meaningful in python, there are strong beliefs about how code should be spaced. take a look at:
https://pypi.python.org/pypi/flake8 (static analysis)
https://www.python.org/dev/peps/pep-0008/ (style enforcement)

my editor runs both for me and highlights in red offending lines, I'm sure there's a way to make them work from your editor. to install them as stand alone tools:
`$ pip install flake8`
`$ pip install pep8`

oh, and now that you've probably installed a few things already via pip, it's probably a good time to tell you that doing it at a global level (`sudo pip install`) pollutes your global env, so you probably want to start using a tool like `virtualenv` + `virtualenvwrapper` soon :) (I can provide more details if necessary, let me know)

I'm going to make a new commit on your branch fixing those problems and we'll continue the review from there

fmezas committed:
Avoid using DOS end of line terminators (^M)

no code changes here. just getting rid of the ^M's

fmezas committed:
Fix pep8/flake8 complaints

fmezas committed:
Refactoring. Extract code into functions

fmezas commented on tetris_3.py:
```diff
@@ -75,6 +54,40 @@ def tetris(d):
+        al -= 1
+
+
+def tetris(pieces_to_drop):
```
extracted most of the code out of this function into smaller ones so that it would be easier to analyse. In general you want to have very short, focused functions that do only one thing and do it right (now that we're talking about that, another interesting reading is https://en.wikipedia.org/wiki/Unix_philosophy ('do one thing and do it well')

also:
https://en.wikipedia.org/wiki/Cohesion_(computer_science)
https://en.wikipedia.org/wiki/Coupling_(computer_programming)

fmezas commented on tetris_3.py:
```diff
@@ -75,6 +54,40 @@ def tetris(d):
+                # if all spaces below the piece are blank, return true
+                row += 1
+            else:
+                array[(row-1)*13 + column] = part
```
I think you've got a problem here. you originally defined `array` as an array (by the way, probably need a better name for that variable, maybe `board`) of 1 char long strings, and here you're assigning a whole piece row (`part`) which potentially is multiple chars long, to a single position in `array`. you end up with potentially a section of the piece in kind of a 'ghost' location and that's why you have to do the `dl` trick in the following lines. you might as well assign the piece chars directly to their right location on the board (`array`)

fmezas commented on tetris_3.py:
```diff
@@ -75,6 +54,40 @@ def tetris(d):
+    al = len(piece_array)  # array length
+    # takes part of a piece individually, not good for Z and S
+    while al >= 1:
+        part = piece_array[al - 1]  # start at the last/bottom part
```
by treating each row of the piece independently of the rest you're allowing them to potentially 'split' as in the second example (problem with the 'S' piece). the way to fix it would be to run this exact same code, except that instead of immediately setting the row of the piece to the found location on the board (line 73 and following), you defer that until you're done with calculating the row for all the rows of the piece, and then just use the one that further up from the bottom (first obstruction found)

fmezas reviewed (commented):
a few comments

fmezas commented on tetris_3.py:
```diff
@@ -27,8 +27,7 @@
 ]
```
by the way, I think I made a mistake here with the definition of the problem. it's not necessary to have nested arrays for the definition of the pieces, it should be just:
```
PIECES = [
  ['HHHH'],
  ['ZZ',
   ' ZZ'],
  [' SS',
   'SS'],
...
]
```
please feel free to change it if it does make sense. my apologies

bberastain committed:
Update tetris_3.py

update_board_with_piece and find_bottom_row both have the same first 8 lines but I couldn't see an easy way to condense that, tired of working on this one, want to move on to the next thing. Please give me any criticism I can apply going forward
